plugins/modules/net_tools/nmcli.py

```diff
@@ -951,6 +972,14 @@ def _compare_conn_params(self, conn_info, options):
 
             if key in conn_info:
                 current_value = conn_info[key]
+                if key == 'ipv4.routes':
+                    # ipv4.routes has not same input and output
+                    # input: ['10.11.0.0/24 10.10.0.2', '10.12.0.0/24 10.10.0.2']
+                    # output: ['{ ip = 10.11.0.0/24', 'nh = 10.10.0.2 }; { ip = 10.12.0.0/24', 'nh = 10.10.0.2 }']
```

Comment:
Shouldn't it be:
```suggestion
                    # ipv4.routes has not same format
                    # value: ['10.11.0.0/24 10.10.0.2', '10.12.0.0/24 10.10.0.2']
                    # current_value: ['{ ip = 10.11.0.0/24', 'nh = 10.10.0.2 }; { ip = 10.12.0.0/24', 'nh = 10.10.0.2 }']
```
?

The code below seems to work on strings, not on lists. So I don't really understand what is happening here. Can you write down how `current_value` looks before and after the code below?

Also, why is this complicated code now necessary, and it wasn't necessary before turning `value` into a list?

Reply:
I understand your POV. Here is what I understood while coding this:

**First point:**

The nmcli command handle most of variables with same input/output. For example, if I wish to set the ipv4.never-default parameters, I will use `ipv4.never-default yes` at command line. And If I ask for current value, I will have the same ouput: `nmcli con NIC show | grep 'ipv4.never-default'` -> `ipv4.never-default:                     yes`

But, with routes, it is different. You set them using this syntax:

`ipv4.routes '10.11.0.0/24 10.10.0.2,10.11.0.0/24 10.10.0.2'`

And when you use the `nmcli show`, the output is:

`ipv4.routes: { ip = 10.11.0.0/24, nh = 10.10.0.2 }; { ip = 10.12.0.0/24, nh = 10.10.0.2 }`

So patterns are not the same.

**Second point:**

The way the nmcli module handle the output is weird. If I already have these 2 routes seen before set, the variable `current_value = conn_info[key]` will not contain a list of 2 item, but 3!! With a very strange pattern:

`['{ ip = 10.11.0.0/24', 'nh = 10.10.0.2 }; { ip = 10.12.0.0/24', 'nh = 10.10.0.2 }']`

So put in YAML list, this is:

```
  - { ip = 10.11.0.0/24
  - nh = 10.10.0.2 }; { ip = 10.12.0.0/24
  - nh = 10.10.0.2 }
```

So compared to the content of `options` variable, which is with same pattern than input, AKA a list like this: `['10.11.0.0/24 10.10.0.2', '10.12.0.0/24 10.10.0.2']`, it is absolutely not the same, and so impossible to compare.

What I did here is merging the output list, then transform it to the input pattern, and using the `current_value = current_value.split(',')` I split it again into a list.

Not sure this is the best method 😐 , I found it to be nice to only modify the value here, so it does not impact anything else in the module.

Reply:
How about adjusting `show_connection()` so it splits this particular list by `;` and not by `,`? Then you have a two-element YAML list
```
  - { ip = 10.11.0.0/24, nh = 10.10.0.2 }
  - { ip = 10.12.0.0/24, nh = 10.10.0.2 }
```
That would be a lot easier to adjust (with a simple regex substitution).

Reply:
Ok, so in the show_connection(), I added a specific split for **ipv4.routes**:

```
        cmd = [self.nmcli_bin, 'con', 'show', self.conn_name]

        (rc, out, err) = self.execute_command(cmd)

        if rc != 0:
            raise NmcliModuleError(err)

        p_enum_value = re.compile(r'^([-]?\d+) \((\w+)\)$')

        conn_info = dict()
        for line in out.splitlines():
            pair = line.split(':', 1)
            key = pair[0].strip()
            key_type = self.settings_type(key)
            if key and len(pair) > 1:
                raw_value = pair[1].lstrip()
                if raw_value == '--':
                    conn_info[key] = None
                elif key == 'bond.options':
                    # Aliases such as 'miimon', 'downdelay' are equivalent to the +bond.options 'option=value' syntax.
                    opts = raw_value.split(',')
                    for opt in opts:
                        alias_pair = opt.split('=', 1)
                        if len(alias_pair) > 1:
                            alias_key = alias_pair[0]
                            alias_value = alias_pair[1]
                            conn_info[alias_key] = alias_value
                elif key == 'ipv4.routes':
                    conn_info[key] = [s.strip() for s in raw_value.split(';')]
                elif key_type == list:
                    conn_info[key] = [s.strip() for s in raw_value.split(',')]
                else:
                    m_enum = p_enum_value.match(raw_value)
                    if m_enum is not None:
                        value = m_enum.group(1)
                    else:
                        value = raw_value
                    conn_info[key] = value
```

I will now replace my for loop in the `_compare_conn_params()` (or should I fully replace it in the `show_connection()` also, so no changes are made in `_compare_conn_params()` ?
What would be for you a "simple regex" in this case ? I still only need to remove all the { , }, ip =, nh = , etc. So at least 4 regex. Note that I am not a Python expert 🙂

Reply:
I'd keep some changes in `_compare_conn_params()`.

About regex: you could create a regex that matches `{ ip = 10.11.0.0/24, nh = 10.10.0.2 }` and converts it to `10.11.0.0/24 10.10.0.2`. Like
```.py
re.sub(r'^{\s*ip\s*=\s*([^, ]+),\s*nh\s*=\s*([^} ]+)\s*}', r'\1 \2', '{ ip = 10.11.0.0/24, nh = 10.10.0.2 }')
```
yields `''10.11.0.0/24 10.10.0.2''`.

Reply:
Impressive. Thanks a lot, it should be ok now.